### PR TITLE
Use weighted random group distribution as default

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst and @havardpe PR

### CODE REVIEW

**Comment by baldersheim** [comment]
Hold merge until a test on one application have been done.

File: searchcore/src/vespa/searchcore/config/partitions.def [hunk 1]
```diff
@@ -50,7 +50,7 @@ dataset[].mpp			int default=1
 dataset[].maxnodesdownperfixedrow int default=0
 
 ## Use simple roundrobin or random.
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/config/partitions.def]: Consider updating comment to reflect new name (weighted)

**Review by geirst — APPROVED** [review]
👍 Only have a minor comment

File: searchcore/src/vespa/searchcore/config/partitions.def (hunk 1 above)
Comment (bratseth) [reply searchcore/src/vespa/searchcore/config/partitions.def]: Agree - it's good if we can change everywhere.

**Review by havardpe — COMMENTED** [review]
👍

**Comment by havardpe** [comment]
I am a bit curious about why we changed the name?
'random'/'round-robin' reflects selection order
'weighted' is more of a selection modifier

What if we want to add weighted hashed selection in the future to improve caching, for example?

**Comment by baldersheim** [comment]
First there was only round-robin.
Then there added random. But it was not good enough. Then it was modified to be weighted random.
So I guess weighted-random is a better name.

BTW, this is ignored if you use grouping or ranking.querycache as that requires hitting the same group as the first request. Then it hashes on the session id.

**Comment by baldersheim** [comment]
After a discussion with @havardpe we concluded that 'adaptive' was better.

**Comment by baldersheim** [comment]
Documentation is fixed in https://github.com/vespa-engine/documentation/pull/276

**Review by geirst — APPROVED** [review]
👍

**Comment by LanceNorskog** [comment]
What has proven more effective than simple random in some situations is:
1) choose 2 possible targets
2) rank targets with your scoring function
3) choose the higher score

**Comment by kkraune** [comment]
@baldersheim is this ready for merge / is it tested on an app?

**Comment by baldersheim** [comment]
This has now been tested and gaven lower and smoother latency. Will now do a more extended test and enable as default after the weekend.
This does not make it default yet.

**Pull request merged**